### DISCUSSION TITLE
ADS-over-MQTT

### DISCUSSION THREAD

**Discussion by tstorek** [stlehmann/pyads#415] [Q&A] [answered]
Thanks for providing the library. It works great so far.

I am wondering if it is also possible to use pyads together with Beckhoff's ADS-over-MQTT.
https://download.beckhoff.com/download/Document/automation/twincat3/ADS-over-MQTT_en.pdf

Unfortunately, I only find examples for setting up the communication from TwinCAT but not from other programming languages.
Hence, I am not sure if it is even possible. Could you maybe enlighten me about this matter.

Thanks, in advance!

Cheers

**Comment by chrisbeardy** [comment, marked as answer]
From reading that document it looks as if   that when you install that special mqtt over ads, it is transparent from the perspective of any end application. Section 4.3. it looks like it just changes the way the ads commands are sent to each router (Using mqtt instead of tcp/IP) , therefore the client app or any pyads connection would still work in the same way so I think pyads would still work. Of course my interpretation of this could be completely wrong as I've not seen this mqtt over ads before.

Personally, I can see a use case for this mqtt over ads implementation but seems very niche and a lot of additional effort over the normal implementation but I see the use case if there is a large network, or encryption is required and Twincat 4024 can't be used. 

If you are just looking for MQTT Comms using Twincat you could use TF6710.

 You would set up an MQTT broker, say mosquitto, then you could communicate to that broker using twincat https://infosys.beckhoff.com/english.php?content=../content/1033/tf6701_tc3_iot_communication_mqtt/18014402002008203.html&id=557070547276860265 and python. 

To use MQTT from python language you need to use a MQTT library,  Something like paho https://pypi.org/project/paho-mqtt/ for python.

There is a good example of this here: https://www.plccoder.com/twincat-and-mqtt-part-1/
